stats/opentelemetry/internal/grpc_trace_bin_propagator_test.go

```diff
@@ -0,0 +1,252 @@
+
+	got := stats.OutgoingTrace(c.Context())
+	want := binary(sc)
+	if string(got) != string(want) {
+		t.Fatalf("got = %v, want %v", got, want)
+	}
+}
+
+// TestInject_SlowPath verifies that the GRPCTraceBinPropagator correctly
+// injects OpenTelemetry trace context data using any other text based carrier.
+//
+// It is called the slow path because it base64 encodes the binary trace
+// context before injecting it.
+func (s) TestInject_SlowPath(t *testing.T) {
+	p := GRPCTraceBinPropagator{}
+	sc := oteltrace.NewSpanContext(oteltrace.SpanContextConfig{
+		TraceID:    [16]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16},
+		SpanID:     [8]byte{17, 18, 19, 20, 21, 22, 23, 24},
+		TraceFlags: oteltrace.FlagsSampled,
+	})
+	tCtx, tCancel := context.WithCancel(context.Background())
+	tCtx = oteltrace.ContextWithSpanContext(tCtx, sc)
+	defer tCancel()
+
+	c := otelpropagation.MapCarrier{}
+	p.Inject(tCtx, c)
+
+	got := c.Get(itracing.GRPCTraceBinHeaderKey)
+	want := base64.StdEncoding.EncodeToString(binary(sc))
+	if got != want {
+		t.Fatalf("got = %v, want %v", got, want)
+	}
+}
+
+// TestExtract_FastPath verifies that the GRPCTraceBinPropagator correctly
+// extracts OpenTelemetry trace context data using the CustomCarrier.
+//
+// It is called the fast path because it extracts the trace context directly
+// in the binary format.
+func (s) TestExtract_FastPath(t *testing.T) {
+	p := GRPCTraceBinPropagator{}
+	sc := oteltrace.NewSpanContext(oteltrace.SpanContextConfig{
+		TraceID:    [16]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16},
+		SpanID:     [8]byte{17, 18, 19, 20, 21, 22, 23, 24},
+		TraceFlags: oteltrace.FlagsSampled,
+		Remote:     true,
+	})
+	bd := binary(sc)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	c := itracing.NewCustomCarrier(stats.SetIncomingTrace(ctx, bd))
+	tCtx := p.Extract(ctx, c)
+	got := oteltrace.SpanContextFromContext(tCtx)
+
+	if !got.Equal(sc) {
+		t.Fatalf("got = %v, want %v", got, sc)
+	}
+}
+
+// TestExtract_SlowPath verifies that the GRPCTraceBinPropagator correctly
+// extracts OpenTelemetry trace context data using any other text based carrier.
+//
+// It is called the slow path because it base64 decodes the binary trace
+// context before extracting it.
+func (s) TestExtract_SlowPath(t *testing.T) {
+	p := GRPCTraceBinPropagator{}
+	sc := oteltrace.NewSpanContext(oteltrace.SpanContextConfig{
+		TraceID:    [16]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16},
+		SpanID:     [8]byte{17, 18, 19, 20, 21, 22, 23, 24},
+		TraceFlags: oteltrace.FlagsSampled,
+		Remote:     true,
+	})
+	bd := binary(sc)
+
+	c := otelpropagation.MapCarrier{
+		itracing.GRPCTraceBinHeaderKey: base64.StdEncoding.EncodeToString(bd),
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	tCtx := p.Extract(ctx, c)
+	got := oteltrace.SpanContextFromContext(tCtx)
+
+	if !got.Equal(sc) {
+		t.Fatalf("got = %v, want %v", got, sc)
+	}
+}
```

Comment:
Let's try and make this a t-test, (I only see the carrier declaration being different).

And then add test cases for all of the failure cases (i.e. all the ones that return just the ctx), you'll be able to verify by whatever oteltrace.SpanContextFromContext returns for an unset span context.

Reply:
yeah merged them under TestExtract and TestInject. Only thing is I am mantaining a boolean `fast` and create the CustomCarrier or other carrier based on that. For inject in case of CustomCarrier, we need to verify if the outgoing trace header was set correctly